lib/index.js

```diff
@@ -262,35 +261,23 @@ Metalsmith.prototype.readFile = unyield(function*(file){
   if (!absolute(file)) file = path.resolve(src, file);
 
   try {
-    var frontmatter = this.frontmatter();
-    var stats = yield fs.stat(file);
-    var buffer = yield fs.readFile(file);
-    var parsed;
-
-    if (frontmatter && utf8(buffer)) {
-      try {
-        parsed = matter(buffer.toString());
-      } catch (e) {
-        var err = new Error('Invalid frontmatter in the file at: ' + file);
-        err.code = 'invalid_frontmatter';
-        throw err;
-      }
-
-      ret = parsed.data;
-      ret.contents = new Buffer(parsed.content);
-    } else {
-      ret.contents = buffer;
-    }
-
-    ret.mode = Mode(stats).toOctal();
-    ret.stats = stats;
+    ret.stats = yield fs.stat(file);
+    ret.contents = yield fs.readFile(file);
+    ret.mode = Mode(ret.stats).toOctal();
   } catch (e) {
-    if (e.code == 'invalid_frontmatter') throw e;
     e.message = 'Failed to read the file at: ' + file + '\n\n' + e.message;
     e.code = 'failed_read';
     throw e;
   }
 
+  if (this.frontmatter()) {
+    // Extract frontmatter
+    var mockFiles = {};
+    mockFiles[file] = ret;
+    frontmatter(mockFiles);
+    ret = mockFiles[file];
+  }
```

Comment:
Note that here I'm interacting with the plugin directly rather than running it via `ware`. I'm not 100% sure whether that was the right choice or not, but it seems clean enough.